[18.0][MIG] account_invoice_refund_option: Migration to 18.0

P-H-Phuc opened the pull request:
This module was initially part of the [Foodcoop set of addons](https://github.com/AwesomeFoodCoops/odoo-production/tree/12.0). We're now migrating it to 18.0 and taking this occasion to contribute it to OCA.

This module enables selecting specific lines for refund.

<img width="980" height="612" alt="image" src="https://github.com/user-attachments/assets/419e9ff6-a2dc-48ce-a666-fe20f93b50bc" />

ajayjaviya committed:
[MIG] Migration to 12.0: account_invoice_refund_option

[IMP] Code

ajayjaviya committed:
[MIG] Migration to 12.0: account_invoice_refund_option

[IMP] Code

druidoo-transbot committed:
[UPD] Update account_invoice_refund_option.pot

druidoo-transbot committed:
[UPD] Update account_invoice_refund_option.pot

ivantodorovich committed:
[IMP] New readme format

[ADD] icon.png

[UPD] README.rst

ivantodorovich committed:
[IMP] New readme format

[ADD] icon.png

[UPD] README.rst

P-H-Phuc committed:
[IMP] account_invoice_refund_option: pre-commit auto fixes

P-H-Phuc committed:
[IMP] account_invoice_refund_option: pre-commit auto fixes

P-H-Phuc committed:
[IMP] account_invoice_refund_option: Migration to 18.0

P-H-Phuc committed:
[IMP] account_invoice_refund_option: Migration to 18.0

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal_view.xml:
```diff
@@ -0,0 +1,40 @@
+            <xpath expr="//field[@name='date']" position="after">
+                <div colspan="4">
+                    <label for="filter_refund" string="Select product(s)" />
+                    <field name="filter_refund" />
```
adjust the alignment of the checkbox `filter_refund`
<img width="483" height="168" alt="image" src="https://github.com/user-attachments/assets/f6a8900b-571b-41b3-80ec-23f9c1556a69" />

nguyenminhchien commented on account_invoice_refund_option/models/account_move.py:
```diff
@@ -0,0 +1,23 @@
+class AccountInvoice(models.Model):
+    _inherit = "account.move"
+
+    def copy_data(self, default=None):
```
why not inherit [def _prepare_default_reversal](https://github.com/odoo/odoo/blob/18.0/addons/account/wizard/account_move_reversal.py#L90) instead

nguyenminhchien reviewed (changes requested):
- Reserve invoice `BILL/2025/09/0006` which has 2 lines:
  + line 1: Product `Storage Box`, qty=10
  + line 2: product `Cabinet with Doors`, qty=10
<img width="1210" height="714" alt="image" src="https://github.com/user-attachments/assets/5c6b0cf3-a2f7-4688-ab16-d3313ecc290e" />

- I reserved line 1, qty=3. It updated `BILL/2025/09/0006` by reducing the qty of `Storage Box` to 3
<img width="1210" height="714" alt="image" src="https://github.com/user-attachments/assets/fde31ccf-6c98-42e8-b5bb-65dfc81f4e99" />

- Expected: Don't update `BILL/2025/09/0006`

nguyenminhchien commented on account_invoice_refund_option/models/account_move.py:
rename to account_move.py

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
rename to account_move_reversal.py

nguyenminhchien commented:
@P-H-Phuc done? remember to request for review after done.

P-H-Phuc commented:
> @P-H-Phuc done? remember to request for review after done.

Not yet. As noted in [comment #1](https://github.com/P-H-Phuc/account-invoicing/pull/3#pullrequestreview-3283285060), it’s a many2many field, so any change on a line will update invoice line

P-H-Phuc commented:
Dear anh @nguyenminhchien,
I’ve made a few updates:

- Introduced a new model `account.move.reversal.refund.line` to allow selecting `products` and `quantity` for reversal.

- Restricted the `Refund Filter option` to single records (form view). The Reverse action will be disabled in multi-record (list view).

- Added a warning when the `reversal quantity` exceeds the `invoiced quantity`.

nguyenminhchien commented:
> Dear anh @nguyenminhchien, I’ve made a few updates:
> 
>     * Introduced a new model `account.move.reversal.refund.line` to allow selecting `products` and `quantity` for reversal.
> 
>     * Restricted the `Refund Filter option` to single records (form view). The Reverse action will be disabled in multi-record (list view).
> 
>     * Added a warning when the `reversal quantity` exceeds the `invoiced quantity`.

is it ready for review? @P-H-Phuc

P-H-Phuc commented:
> > Dear anh @nguyenminhchien, I’ve made a few updates:
> > ```
> > * Introduced a new model `account.move.reversal.refund.line` to allow selecting `products` and `quantity` for reversal.
> > 
> > * Restricted the `Refund Filter option` to single records (form view). The Reverse action will be disabled in multi-record (list view).
> > 
> > * Added a warning when the `reversal quantity` exceeds the `invoiced quantity`.
> > ```
> 
> is it ready for review? @P-H-Phuc

Yes, it's already

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
```diff
@@ -0,0 +1,113 @@
+            self.refund_line_ids = []
+
+    def _prepare_default_reversal(self, move):
+        move = self.env["account.move"].browse(move.id)
```
`move` is a recordset aleady, why do you call `browse(move.id) `?

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
```diff
@@ -0,0 +1,113 @@
+        return vals
+
+
+class AccountMoveReversalRefundLine(models.TransientModel):
```
create new py file for this new class

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
```diff
@@ -0,0 +1,113 @@
+    def _compute_available_product_ids(self):
+        for record in self:
+            product_ids = record.move_ids.line_ids.mapped("product_id")
+            record.available_product_ids = product_ids or False
```
```suggestion
            record.available_product_ids = record.mapped("move_ids.line_ids.product_id")
```

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
```diff
@@ -0,0 +1,113 @@
+                ],
+                limit=1,
+            )
+            record.quantity = sum(line.mapped("quantity"))
```
You search with limit=1, no need to sum.

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal.py:
```diff
@@ -0,0 +1,113 @@
+            default=0.0,
+        )
+        if self.quantity > max_qty_refund:
+            self.quantity = max_qty_refund
```
Don't set the value automatically, it could lead to the confusion for the user.
Anw, you should check the constraint for qty instead. I mean, add decorator `@api.constrains` instead of `@api.onchange`

nguyenminhchien reviewed (changes requested):
It's weird behavior when clicking on button `Reserve and Create Invoice`. hmm, we don't have it in version 12.0, you can hide this button when `Select product(s)` ticked.

<img width="980" height="622" alt="image" src="https://github.com/user-attachments/assets/16119884-ab43-46f5-aba2-a0c3cf2e6e17" />

P-H-Phuc commented:
All done!

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal_refund_line.py:
```diff
@@ -0,0 +1,69 @@
+                ],
+                limit=1,
+            )
+            record.quantity = line.quantity if line else 0.0
```
```suggestion
            record.quantity = line.quantity
```

nguyenminhchien commented on account_invoice_refund_option/wizard/account_move_reversal_refund_line.py:
```diff
@@ -0,0 +1,69 @@
+            if record.quantity > max_qty_refund:
+                raise ValidationError(
+                    self.env._(
+                        "The maximum quantity to refund for the product %s is %s.",
```
```suggestion
                        "The refund quantity for [Product Name] must be max_qty_refund or less. Please reduce the quantity and try again.",
```
